### PR TITLE
E2e protocol ricardo

### PR DESCRIPTION
**Pull request by corrideat**

### CODE REVIEW

File: frontend/controller/actions/chatroom.js [hunk 1]
```diff
@@ -121,7 +129,17 @@ export default (sbp('sbp/selectors/register', {
             foreignKey: cekOpts.foreignKey,
             meta: cekOpts.meta,
             data: cekOpts.data
-          }
+          },
+          encryptedOutgoingDataWithRawKey(CEK, {
```
Comment (taoeffect) [inline frontend/controller/actions/chatroom.js:134]: Can you leave a TODO comment here to figure out how to avoid having the developer interact with `encryptedOutgoingDataWithRawKey`?

File: frontend/controller/actions/chatroom.js [hunk 2]
```diff
@@ -174,8 +192,39 @@ export default (sbp('sbp/selectors/register', {
...
+    const CEKid = sbp('chelonia/contract/currentKeyIdByName', params.contractID, 'cek')
+    const userCSKid = sbp('chelonia/contract/currentKeyIdByName', userID, 'csk')
+
+    const state = rootState[params.contractID]
+
+    // Add the user's CSK to the contract
+    await sbp('chelonia/out/keyAdd', {
+      contractID: params.contractID,
+      contractName: 'gi.contracts/chatroom',
+      data: [encryptedOutgoingData(state, CEKid, {
```
Comment (taoeffect) [inline frontend/controller/actions/chatroom.js:217]: Same request here re TODO comment

File: frontend/controller/actions/group.js [hunk 3]
```diff
@@ -213,6 +220,18 @@ export default (sbp('sbp/selectors/register', {
...
+           encryptedOutgoingDataWithRawKey(CEK, {
+            foreignKey: `sp:${encodeURIComponent(userID)}?keyName=${encodeURIComponent('csk')}`,
+            id: userCSKid,
+            data: rootState[userID]._vm.authorizedKeys[userCSKid].data,
+            permissions: [GIMessage.OP_ACTION_ENCRYPTED + '#inner'],
+            allowedActions: '*',
+            purpose: ['sig'],
+            ringLevel: Number.MAX_SAFE_INTEGER,
+            name: `${userID}/${userCSKid}`
+          }) */
```
Comment (taoeffect) [inline frontend/controller/actions/group.js]: If this is no longer needed can it be removed?

File: frontend/controller/actions/group.js [hunk 4]
```diff
@@ -281,11 +301,24 @@ export default (sbp('sbp/selectors/register', {
           }
         },
         signingKeyId: CSKid,
-        encryptionKeyId: CEKid
+        encryptionKeyId: CEKid,
+        innerSigningContractID: null
```
Comment (taoeffect) [inline frontend/controller/actions/group.js:297]: Does this need to be specified as `null` or can it be removed?

If it needs to be specified as `null` can you add a comment explaining why?

File: frontend/controller/actions/group.js [hunk 5]
```diff
@@ -281,11 +301,24 @@ export default (sbp('sbp/selectors/register', {
...
+      await sbp('gi.actions/group/joinAndSwitch', {
+        originatingContractID: userID,
+        originatingContractName: 'gi.contracts/identity',
+        contractID: contractID,
+        contractName: 'gi.contracts/group',
+        signingKeyId: CSKid,
+        innerSigningKeyId: userCSKid,
+        encryptionKeyId: userCEKid,
+        options: {
+          skipUsableKeysCheck: true
```
Comment (taoeffect) [inline frontend/controller/actions/group.js:313]: Can you please add a comment explaining why this is being set to `true` here?

File: frontend/controller/actions/group.js [hunk 6]
```diff
@@ -309,27 +342,28 @@ export default (sbp('sbp/selectors/register', {
...
+      keyIds.length && await sbp('gi.actions/out/shareVolatileKeys', {
         contractID: contractID,
         contractName: 'gi.contracts/group',
         subjectContractID: userID,
-        keyIds: [PEKid, ...revokedPEKids]
+        keyIds: keyIds
       })
 
+      */
+
```
Comment (taoeffect) [inline frontend/controller/actions/group.js]: If all this code is no longer needed should it be deleted?

File: frontend/model/contracts/group.js [hunk 7]
```diff
@@ -552,9 +552,7 @@ sbp('chelonia/defineContract', {
             allowPublicChannels: false
           },
           streaks: initGroupStreaks(),
-          profiles: {
-            [meta.username]: initGroupProfile(meta.createdDate)
-          },
+          profiles: {},
```
Comment (taoeffect) [inline frontend/model/contracts/group.js:555]: Why was this removed?

File: shared/domains/chelonia/chelonia.js [hunk 8]
```diff
@@ -567,6 +565,26 @@ export default (sbp('sbp/selectors/register', {
     // calling this will make pubsub unsubscribe for events on `contractID`
     sbp('okTurtles.events/emit', CONTRACTS_MODIFIED, state.contracts)
   },
+  'chelonia/contract/disconnect': async function (contractID, contractIDToDisconnect) {
```
Comment (taoeffect) [inline shared/domains/chelonia/chelonia.js:568]: Is there a corresponding `'chelonia/contract/connect'`? If not, why not? It seems like there should be one for clarity.

File: frontend/model/contracts/group.js (hunk 7 above)
Comment (corrideat) [reply frontend/model/contracts/group.js:555]: Because it's set when joining the group, which is now something that the group creator also does. (Also, because `meta.username` will be removed).

File: shared/domains/chelonia/chelonia.js [hunk 9]
```diff
@@ -567,6 +565,26 @@ export default (sbp('sbp/selectors/register', {
...
+    const state = sbp(this.config.stateSelector)
+    const contractState = state[contractID]
+
+    const keyIds = Object.values(contractState._vm.authorizedKeys).filter((k) => {
+      // $FlowFixMe
+      return k._notAfterHeight == null && k.meta?.keyRequest?.contractID === contractIDToDisconnect
+    }).map(k => (k: any).id)
+
+    console.log('chelonia/contract/disconnect:  ', { keyIds })
```
Comment (taoeffect) [inline shared/domains/chelonia/chelonia.js]: Is this needed? Can it be removed? (We should start cleaning up the excess logging...)

File: shared/domains/chelonia/chelonia.js [hunk 10]
```diff
@@ -567,6 +565,26 @@ export default (sbp('sbp/selectors/register', {
...
+
+    if (!keyIds.length) return
```
Comment (taoeffect) [inline shared/domains/chelonia/chelonia.js:577]: Is this a silent error? (All errors should be logged, exceptions thrown, etc.)

File: frontend/controller/actions/group.js (hunk 6 above)
Comment (corrideat) [reply frontend/controller/actions/group.js]: This is related to how the group creator also joins normally (and it can be deleted).

File: shared/domains/chelonia/GIMessage.js [hunk 11]
```diff
@@ -50,61 +62,143 @@ export type GIOpKeyUpdate = {
...
+            if (computedKeyId !== key.id) {
+              throw new Error(`Key ID mismatch. Expected to decrypt key ID ${key.id} but got ${computedKeyId}`)
+            }
+          })
+        }
+        if (key.meta?.keyRequest?.contractID) {
+          try {
+            key.meta.keyRequest.contractID = maybeEncryptedIncomingData(contractID, state, key.meta.keyRequest.contractID, height, additionalKeys, headJSON)?.valueOf()
+          } catch {
+            delete key.meta.keyRequest.contractID
```
Comment (taoeffect) [inline shared/domains/chelonia/GIMessage.js:110]: Silent error? If an error shouldn't be logged, a comment should always explain why not.

Otherwise, errors should always be logged.

File: frontend/controller/actions/group.js (hunk 4 above)
Comment (corrideat) [reply frontend/controller/actions/group.js:297]: Maybe it doesn't need to be `null` but it makes more sense, since the General chatroom is part of creating the group and not created by any specific user (technically, it's created by the group creator, but that's an implementation detail).

File: shared/domains/chelonia/chelonia.js (hunk 8 above)
Comment (corrideat) [reply shared/domains/chelonia/chelonia.js:568]: There isn't by that name, but it'd be `/out/keyRequest'.

File: frontend/controller/actions/group.js (hunk 3 above)
Comment (corrideat) [reply frontend/controller/actions/group.js]: It can

File: shared/domains/chelonia/GIMessage.js [hunk 12]
```diff
@@ -50,61 +62,143 @@ export type GIOpKeyUpdate = {
...
+  // If the operation is GIMessage.OP_KEY_SHARE,
+  // extract encrypted data from keys?.[].meta?.private?.content
+  if (op === GIMessage.OP_KEY_SHARE) {
+    return maybeEncryptedIncomingData(contractID, state, (message: any), height, additionalKeys, headJSON, (message) => {
+      (message: any).keys?.forEach((key) => {
+        if (!key.meta?.private?.content) return
         const decryptionFn = message.foreignContractID ? encryptedIncomingForeignData : encryptedIncomingData
-        const decryptionContract = message.foreignContractID ? message.foreignContractID : contractID
-        key.meta.private.content = decryptionFn(decryptionContract, state, key.meta.private.content, height, additionalKeys, (value) => {
+        const decryptionContract = String(message.foreignContractID ? message.foreignContractID : contractID)
```
Comment (taoeffect) [inline shared/domains/chelonia/GIMessage.js]: Can't this be simplified to:

```js
message.foreignContractID || contractID
```

without the `String` wrapper?

Comment (corrideat) [reply shared/domains/chelonia/GIMessage.js]: The `String` wrapper is for casting as it needs to be a string.

File: shared/domains/chelonia/chelonia.js [hunk 13]
```diff
@@ -852,64 +893,67 @@ export default (sbp('sbp/selectors/register', {
...
+    const signingKeyId = findSuitableSecretKeyId(originatingState, [GIMessage.OP_KEY_ADD], ['sig'])
+    if (!signingKeyId) {
+      throw ChelErrorUnexpected(`Unable to send key request. Originating contract is missing a key with OP_KEY_ADD permission. contractID=${contractID} originatingContractID=${originatingContractID}`)
+    }
+    await sbp('chelonia/out/keyAdd', {
+      contractID: originatingContractID,
+      contractName: originatingContractName,
+      data: [{
+        id: keyRequestReplyKeyId,
+        name: 'krrk-' + keyRequestReplyKeyId,
```
Comment (taoeffect) [inline shared/domains/chelonia/chelonia.js]: Either needs to be done now or later, all internal key names (and reserved keys) should be prefixed with `#`

File: shared/domains/chelonia/crypto.js [hunk 14]
```diff
@@ -255,7 +255,7 @@ export const verifySignature = (inKey: Key | string, data: string, signature: st
     throw new Error('Invalid signature')
   }
 }
-export const encrypt = (inKey: Key | string, data: string): string => {
+export const encrypt = (inKey: Key | string, data: string, ad?: string): string => {
```
Comment (taoeffect) [inline shared/domains/chelonia/crypto.js:264]: What is `ad`? Can you please add comments here explaining function parameters, especially `ad` which is very unclear?

File: shared/domains/chelonia/chelonia.js (hunk 9 above)
Comment (corrideat) [reply shared/domains/chelonia/chelonia.js]: Not needed other than for debugging purposes

File: shared/domains/chelonia/internals.js [hunk 15]
```diff
@@ -159,6 +196,13 @@ export default (sbp('sbp/selectors/register', {
     saferEval({
       // pass in globals that we want access to by default in the sandbox
       // note: you can undefine these by setting them to undefined in exposedGlobals
+      ...(typeof self !== 'undefined' && {
+        self: {
+          crypto: {
+            getRandomValues: (v) => self.crypto.getRandomValues(v)
+          }
+        }
+      }),
```
Comment (taoeffect) [inline shared/domains/chelonia/internals.js]: Should this be exposed under `self`...? None of the other globals are. Maybe it should be directly included under `crypto`?

(And are we sure this is necessary here still?)

File: shared/domains/chelonia/internals.js [hunk 16]
```diff
@@ -881,19 +930,55 @@ export default (sbp('sbp/selectors/register', {
...
+            }))
           },
-          signingKeyId
+          signingKey: deserializedResponseKey
         })
 
         // 4(i). Remove originating contract and update current contract with information
-        await sbp('chelonia/out/keyRequestResponse', { contractID, contractName, signingKeyId, data: { keyRequestHash: hash, success: true } })
+        const payload = { keyRequestHash: hash, success: true }
+        // TODO: Combine these two operations (KRR and KS) into one using OP_ATOMIC
```
Comment (taoeffect) [inline shared/domains/chelonia/internals.js]: Is this something that can be done now? Or should we do it in a separate PR?

File: shared/domains/chelonia/chelonia.js (hunk 10 above)
Comment (corrideat) [reply shared/domains/chelonia/chelonia.js:577]: This isn't an error per se. If you call disconnect but there is nothing to disconnect, it returns (this ensures idempotency).

**Review by taoeffect — CHANGES_REQUESTED** [review]
Great work @corrideat!! There are some flow errors, those need to be fixed before merging. Also see other comments

**Review by taoeffect — APPROVED** [review]
LGTM! Nice job @corrideat !!
